lib/srv/authhandlers.go

```diff
@@ -151,88 +149,58 @@ func NewAuthHandlers(config *AuthHandlerConfig) (*AuthHandlers, error) {
 // CreateIdentityContext returns an IdentityContext populated with information
 // about the logged in user on the connection.
 func (h *AuthHandlers) CreateIdentityContext(sconn *ssh.ServerConn) (IdentityContext, error) {
-	identity := IdentityContext{
-		TeleportUser: sconn.Permissions.Extensions[utils.CertTeleportUser],
-		Login:        sconn.User(),
-	}
-
-	clusterName, err := h.c.AccessPoint.GetClusterName()
-	if err != nil {
-		return IdentityContext{}, trace.Wrap(err)
-	}
-
 	certRaw := []byte(sconn.Permissions.Extensions[utils.CertTeleportUserCertificate])
 	certificate, err := apisshutils.ParseCertificate(certRaw)
 	if err != nil {
 		return IdentityContext{}, trace.Wrap(err)
 	}
-	identity.Certificate = certificate
-	identity.RouteToCluster = certificate.Extensions[teleport.CertExtensionTeleportRouteToCluster]
-	if certificate.ValidBefore != 0 {
-		identity.CertValidBefore = time.Unix(int64(certificate.ValidBefore), 0)
-	}
-	certAuthority, err := h.authorityForCert(types.UserCA, certificate.SignatureKey)
+
+	unmappedIdentity, err := sshca.DecodeIdentity(certificate)
 	if err != nil {
 		return IdentityContext{}, trace.Wrap(err)
 	}
-	identity.CertAuthority = certAuthority
 
-	identity.UnmappedRoles, err = services.ExtractRolesFromCert(certificate)
-	if err != nil {
-		return IdentityContext{}, trace.Wrap(err)
+	ictx := IdentityContext{
+		UnmappedIdentity:        unmappedIdentity,
+		Login:                   sconn.User(),
+		TeleportUser:            unmappedIdentity.Username,
+		RouteToCluster:          unmappedIdentity.RouteToCluster,
+		UnmappedRoles:           unmappedIdentity.Roles,
+		AllowedResourceIDs:      unmappedIdentity.AllowedResourceIDs,
+		Impersonator:            unmappedIdentity.Impersonator,
+		ActiveRequests:          unmappedIdentity.ActiveRequests,
+		DisallowReissue:         unmappedIdentity.DisallowReissue,
+		Renewable:               unmappedIdentity.Renewable,
+		BotName:                 unmappedIdentity.BotName,
+		BotInstanceID:           unmappedIdentity.BotInstanceID,
+		Generation:              unmappedIdentity.Generation,
+		PreviousIdentityExpires: unmappedIdentity.PreviousIdentityExpires,
 	}
 
-	accessInfo, err := fetchAccessInfo(certificate, certAuthority, identity.TeleportUser, clusterName.GetClusterName())
+	clusterName, err := h.c.AccessPoint.GetClusterName()
 	if err != nil {
 		return IdentityContext{}, trace.Wrap(err)
 	}
-	identity.AllowedResourceIDs = accessInfo.AllowedResourceIDs
-	identity.AccessChecker, err = services.NewAccessChecker(accessInfo, clusterName.GetClusterName(), h.c.AccessPoint)
+
+	if unmappedIdentity.ValidBefore != 0 {
+		ictx.CertValidBefore = time.Unix(int64(unmappedIdentity.ValidBefore), 0)
+	}
```

Comment:
Suggestion: add a space? This seems unrelated from the rest.

```suggestion
	}

```